### PR TITLE
Consolidate image & executable build into exe-in-image

### PR DESCRIPTION
**Pull request by mpozniak95**
## Related Issue
https://github.com/Granulate/gprofiler/issues/582

### CODE REVIEW

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:772]: Bring back :)

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:787]: I think you can just check `if args.subcommand == extract-resources` then call `copy_resources` then `return`. This way, we don't need to make any other changes in the `main` function

File: tests/test.sh
Comment (Jongy) [inline tests/test.sh:33]:
```suggestion
    echo "Please run \"gprofiler extract-resources --resources-dest=$SCRIPT_DIR/../gprofiler/resources/\" to get all resources"
```

File: tests/test_sanity.py
Comment (Jongy) [inline tests/test_sanity.py:201]:
```suggestion
            " with phpspy grep: https://github.com/Granulate/gprofiler/issues/794"
```

File: pyi.Dockerfile
Comment (Jongy) [inline pyi.Dockerfile:118]: No need to modify the staticx config for pyperf  - the idea is to make gprofiler build faster when modifying only python code. I don't want to change the build logic of pyperf itself, only of the pyinstaller of gprofiler.

File: Dockerfile
Comment (Jongy) [inline Dockerfile:219]: Please keep tis comment over `GPROFILER_IN_CONTAINER`.

**Review by Jongy — CHANGES_REQUESTED** [review]
I reviewed partially. i still need to go over the CI changes and ensure that we didn't miss anything in the removed `Dockerfile`.

File: .dockerignore
Comment (mpozniak95) [reply .dockerignore:2]: Executable that is copied into docker image is located in build dir, thats why it is not ignored anymore

File: gprofiler/main.py
Comment (mpozniak95) [reply gprofiler/main.py:772]: It was never deleted, not sure why git shows this deletion here :o

**Review by marcin-ol — COMMENTED** [review]
Reviewed the changes, compared  `pyi.Dockerfile` with `Dockerfile` from master, according to suggestion from @Jongy:
- Found no key parts missing from Dockerfile (former container-build), step-wise.
- Verified resources referenced in `build-stage` with final layer of container build.

No remarks wrt build on x86.

**Comment by marcin-ol** [comment]
Verified executable build and test on aarch64.
Executable tests complete without issues.

**Comment by marcin-ol** [comment]
Building and running container tests on aarch64 results in error:
```
E           docker.errors.ContainerError: Command '['-v', '-d', '3', '-o', '/tmp/gprofiler', '-d', '3', '--no-java', '--no-php', '--no-ruby', '--no-nodejs', '--no-dotnet', '--python-mode=pyperf']' in image '<Image: 'gprofiler:latest'>' returned non-zero exit status 1: b'

exec /gprofiler: no such file or directory\n'
```

Seems that gprofiler bootstrap in container needs attention:
```
/ # ldd /gprofiler
        /lib/ld-linux-aarch64.so.1 (0xffffa7d5d000)
        libdl.so.2 => /lib/ld-linux-aarch64.so.1 (0xffffa7d5d000)
        libz.so.1 => /lib/libz.so.1 (0xffffa7d2c000)
        libpthread.so.0 => /lib/ld-linux-aarch64.so.1 (0xffffa7d5d000)
        libc.so.6 => /lib/ld-linux-aarch64.so.1 (0xffffa7d5d000)
Error relocating /gprofiler: __realpath_chk: symbol not found
Error relocating /gprofiler: __strdup: symbol not found
```

File: pyi.Dockerfile
Comment (marcin-ol) [inline pyi.Dockerfile:296]:
```suggestion
    if [ "$STATICX" = "true" ]; then \
```
`staticx` should be run on aarch64 for gprofiler to ensure correct loading.

Comment (Jongy) [reply pyi.Dockerfile]: Yup, it should :( we broke it. It is fixed in https://github.com/Granulate/gprofiler/pull/806

File: gprofiler/main.py
Comment (Jongy) [reply gprofiler/main.py:772]: In master it looks like:
```
    args = parser.parse_args()

    args.perf_inject = args.nodejs_mode == "perf"
    args.perf_node_attach = args.nodejs_mode == "attach-maps"
```

how does it look in your branch?

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:1094]:
```suggestion
            if args.subcommand == "extract-resources":
                args.func(args.resources_dest)
            else:
                assert args.subcommand == "upload-file"
                args.func(args, profiler_api_client)
            return
```

File: pyi.Dockerfile
Comment (Jongy) [reply pyi.Dockerfile:296]: Or actually, yeah, only if not `--fast`.

File: gprofiler/main.py
Comment (mpozniak95) [reply gprofiler/main.py]: Oh, ok. I changed it back.

**Review by Jongy — COMMENTED** [review]
Getting this build error:
```
 > [bcc-build 14/14] RUN ./pyperf_build.sh --with-staticx:
#59 0.203 Cloning into 'bcc'...
#59 7.473 fatal: Could not parse object '598855b0e297e659f4f880a6fcf49dc173eb5631'.
```

Which is fixed once you merge master, meanwhile I did it locally as I don't have push permissions to your repo

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:653]: `None` is a saner sentinel - so default to `None` and check `is not None`.

File: gprofiler/main.py
Comment (Jongy) [inline gprofiler/main.py:1098]: `extract-resources`, unlike `upload-file` is a subcommand which requires much less preparation/verification of preconditions. By putting its handling this late in the code, you have to fix various sites like
```
    if not args.upload_results and not args.output_dir and not args.extract_resources:
        parser.error("Must pass at least one output method (--upload-results / --output-dir)")
```
and there are also some unfixed sites - for example, `verify_preconditions` still runs and unnecessarily checks for root, etc.

Please move `extract-resources` handling to be earlier, right after `parse_args()`. Also don't forget to move this check:
```
    if args.extract_resources and args.resources_dest == "":
        parser.error("Must provide --resources-dest when extract-resources")
```

File: Dockerfile
Comment (Jongy) [inline Dockerfile:8]: Add newline before EOF (so after line 8)

File: Dockerfile
Comment (Jongy) [inline Dockerfile:1]: Some spacing would make in this file to make it readable. between the `FROM` to next lines, before `ENTRYPOINT`.

File: Dockerfile
Comment (Jongy) [inline Dockerfile:15]: Please move this comment to its appropriate location in `build_x86_64_executable.sh`, I don't want this information forgotten.

File: Dockerfile
Comment (Jongy) [inline Dockerfile:1]: On this same occasion, let's rename this file to `container.Dockerfile` and fix related build scripts.

File: pyi.Dockerfile
Comment (Jongy) [inline pyi.Dockerfile:1]: Continuing my suggestion about `container.Dockerfile`, this should be `executable.Dockerfile`.

File: tests/test_app_metadata.py
Comment (Jongy) [inline tests/test_app_metadata.py:195]: So by switching the base layer of the container to alpine, we break .NET completely for the container build? (not saying it's a problem, as .NET hasn't been ours/users' focus lately, just trying to understand)

File: scripts/build_aarch64_container.sh
Comment (Jongy) [inline scripts/build_aarch64_container.sh:1]: To make sure this one (and `build_x86_64_container.sh`) always use a fresh build of the executable, and not a stale result - please invoke the matching executable build script here before rebuilding the image itself.

**Review by Jongy — COMMENTED** [review]
Got this failure trying to build in `--fast` mode:
```
 => [build-stage 41/41] RUN set -e;     if [ "false" = "true" ]; then         LIBS=$(./scripts/list_needed_libs.sh) &&         staticx $LIBS dist/gprofiler dist/gprofiler.static ;      0.3s
 => ERROR [export-stage 1/1] COPY --from=build-stage /app/dist/gprofiler.static /gprofiler                                                                                               0.0s
------
 > [export-stage 1/1] COPY --from=build-stage /app/dist/gprofiler.static /gprofiler:
------
failed to compute cache key: "/app/dist/gprofiler.static" not found: not found
```

If `--fast` you need to copy `/app/dist/gprofiler`, not the `.static` one

File: .github/workflows/build-executable-test-deploy.yml
Comment (Jongy) [inline .github/workflows/build-executable-test-deploy.yml:1]: This file has gotten very long and condensed. I'm not sure there's a way to split it to separate workflows, because all jobs here are dependent in some way. Please spend some time thinking if this can be split somehow.

If we don't split - let's at least give this file a shorter filename (and `name:` in the yaml). `build-test-deploy.yml` and `name: Build, test and deploy`

File: .github/workflows/build-executable-test-deploy.yml
Comment (Jongy) [inline .github/workflows/build-executable-test-deploy.yml:255]: Isn't this job & `build-container-x64` the same? This seems like some duplication.

File: .github/workflows/build-executable-test-deploy.yml
Comment (Jongy) [reply .github/workflows/build-executable-test-deploy.yml:1]: Also, if we can't split, please document in a comment in the top of this file WHY we can't split.

Comment (Jongy) [inline .github/workflows/build-executable-test-deploy.yml:1]: As the file has many jobs now, I suggest we maintain a stricter naming convention to make things clear.

Let's have these job names (I'm using shell patterns to convey permutations)
* build-{executable,container}-{x64,aarch64} - no "build-docker".
* test-{executable,container-and-unit}-x64 - add x64 as a suffix, since all tests right now are "x64". @pfilipko1 is working on aarch64 tests in https://github.com/Granulate/gprofiler/pull/822, I believe we will first merge this PR and then @pfilipko1 will add the -aarch64 jobs.
* push-manifest should be renamed to deploy-container, to match deploy-executable. Both don't require arch suffix, because they deploy both arches.

**Review by Jongy — CHANGES_REQUESTED** [review]
I finished reviewing everything :)

Besides CI that runs on pull request, I'd want to test the new deployment jobs after all fixes. When you've fixed everything please coordinate with me a test. We'll need to edit some stuff to avoid pushing the `latest` image.

File: tests/test_app_metadata.py
Comment (mpozniak95) [reply tests/test_app_metadata.py:196]: Unfortunately yes :(

File: .github/workflows/build-executable-test-deploy.yml
Comment (mpozniak95) [reply .github/workflows/build-executable-test-deploy.yml]: Some steps of job `build-docker-x64` were the same as steps in `build-container-x64`, and I moved other steps from that job to `deploy-container` job

File: executable.Dockerfile
Comment (Jongy) [inline executable.Dockerfile:298]:
```suggestion
        staticx $LIBS dist/gprofiler dist/gprofiler.output ; \  
```

**Comment by Jongy** [comment]
I finished re-reviewing all past comments, I want to have one last go at the CI file and we'll be able to merge it.

**Review by Jongy — APPROVED** [review]
Good to merge. I have some final nitpicks that I'll handle in a separate PR after this.

Good job @mpozniak95 , this is a core change with many different aspects and subtleties and you've handled all of them thoroughly and properly :)

**Pull request merged**
